File: rearrangingFruits.js
/**
 * @param {number[]} basket1
 * @param {number[]} basket2
 * @return {number}
 */
var minCost = function(basket1, basket2) {
    const freq = new Map();
    let minFruit = Infinity;

    // Count net frequency difference and track global minimum fruit cost
    for (let i = 0; i < basket1.length; i++) {
        const a = basket1[i], b = basket2[i];
        freq.set(a, (freq.get(a) || 0) + 1);
        freq.set(b, (freq.get(b) || 0) - 1);
        minFruit = Math.min(minFruit, a, b);
    }

    const excess = [];

    for (const [val, count] of freq.entries()) {
        if (count % 2 !== 0) return -1;
        const half = Math.abs(count) / 2;
        for (let i = 0; i < half; i++) excess.push(val);
    }

    if (excess.length === 0) return 0;

    excess.sort((a, b) => a - b);

    let cost = 0;
    const swaps = excess.length / 2;
    for (let i = 0; i < swaps; i++) {
        cost += Math.min(excess[i], 2 * minFruit);
    }

    return cost;
};

File: rearrangingFruits.cs
using System;
using System.Collections.Generic;

public class Solution {
    public long MinCost(int[] basket1, int[] basket2) {
        var freq = new Dictionary<int, int>();
        int minFruit = int.MaxValue;

        // Count net frequency difference
        for (int i = 0; i < basket1.Length; i++) {
            int a = basket1[i], b = basket2[i];
            freq[a] = freq.ContainsKey(a) ? freq[a] + 1 : 1;
            freq[b] = freq.ContainsKey(b) ? freq[b] - 1 : -1;
            minFruit = Math.Min(minFruit, Math.Min(a, b));
        }

        var excess = new List<int>();

        foreach (var kvp in freq) {
            int val = kvp.Key, count = kvp.Value;
            if (count % 2 != 0) return -1;
            int half = Math.Abs(count) / 2;
            for (int i = 0; i < half; i++) excess.Add(val);
        }

        excess.Sort();

        long cost = 0;
        int m = excess.Count;
        for (int i = 0; i < m / 2; i++) {
            cost += Math.Min(excess[i], 2 * minFruit);
        }

        return cost;
    }
}
